ingestors/kafka/src/main/scala/hydra/kafka/util/ConsumerGroupsOffsetConsumer.scala

```diff
@@ -0,0 +1,223 @@
+    } yield ()
+  }
+
+  // Gets partition info for the kafkaInternalTopic "__consumer_offsets" and initializes each to offset to 0
+  private def initializePartitions[F[_]: ConcurrentEffect](kafkaAdminAlgebra: KafkaAdminAlgebra[F], kafkaInternalTopic: String): F[Ref[F, PartitionOffsetMap]] = {
+    for {
+      consumerOffsetsLatestOffsets <- kafkaAdminAlgebra.getLatestOffsets(kafkaInternalTopic).map { m =>
+        m.map(p => (p._1.partition, 0L))
+      }
+      cache <- Ref[F].of(consumerOffsetsLatestOffsets)
+    } yield cache
+  }
+
+  private def getConsumerGroupDeserializer[F[_]: Sync, A](byteBufferToA: ByteBuffer => A): Deserializer[F, Option[A]] =
+    Deserializer.delegate[F, Option[A]] {
+      (_: String, data: Array[Byte]) => {
+        Try(byteBufferToA(Bytes(data).toByteBuffer)).toOption
+      }
+    }.suspend
+
+  private def getSerializer[F[_]: Sync, A](schemaRegistryClient: SchemaRegistryClient)(isKey: Boolean): Serializer[F, A] =
+    Serializer.delegate[F, A] {
+      val serializer = {
+        val se = new KafkaAvroSerializer(schemaRegistryClient)
+        se.configure(Map(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG -> "").asJava, isKey)
+        se
+      }
+      (topic: String, data: A) => serializer.serialize(topic, data)
+    }.suspend
+
+  private def processRecord[F[_]: ConcurrentEffect: Logger](
+                                                             cr: CommittableConsumerRecord[F, Option[BaseKey], Option[OffsetAndMetadata]],
+                                                             s: SchemaRegistryClient,
+                                                             destinationTopic: String,
+                                                             dvsInternalKafkaOffsetTopic: String
+                                                           ): fs2.Stream[F, ProducerRecords[Array[Byte], Array[Byte], Unit]] = {
+    ((cr.record.key, cr.record.value) match {
+      case (Some(OffsetKey(_, k)), offsetMaybe) =>
+        val topicMaybe: Option[String] = Option(k.topicPartition.topic())
+        val consumerGroupMaybe: Option[String] = Option(k.group)
+        val consumerKeyMaybe: Option[TopicConsumerKey] = consumerGroupMaybe.flatMap(cg => topicMaybe.map(t => TopicConsumerKey(t, cg)))
+        val consumerValue = offsetMaybe.map(o => Instant.ofEpochMilli(o.commitTimestamp)).map(TopicConsumerValue.apply)
+
+        consumerKeyMaybe match {
+          case Some(consumerKey) =>
+            fs2.Stream.eval(for {
+              topicConsumer <- TopicConsumer.encode[F](consumerKey, consumerValue)
+              topicConsumerOffset <- TopicConsumerOffset.encode[F](
+                TopicConsumerOffsetKey(cr.offset.topicPartition.topic(),cr.offset.topicPartition.partition()),
```

Comment:
Is the CommittableConsumerRecord topic different than the topicMaybe? Or do they point to the same topic?

Reply:
The CommittableRecord gives us the data for our offsets-offsets topic. The other gives us the data for the cache.